[FSDP] Relax post-backward assert

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89791

This assert was accidentally made stricter when transitioning from per-FSDP-instance training state to per-handle training state. This PR relaxes it again, which should restore compatibility for some reentrant AC plus FSDP cases.

Commit:
[FSDP] Relax post-backward assert

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -482,9 +482,13 @@ def _post_backward_hook(
         "FullyShardedDataParallel._post_backward_hook"
     ):
         _assert_in_training_states(state, [TrainingState.FORWARD_BACKWARD])
+        # For reentrant AC, the post-backward hook may run multiple times in
```
nit: For reentrant AC multiple times

Commit:
Update on "[FSDP] Relax post-backward assert"


This assert was accidentally made stricter when transitioning from per-FSDP-instance training state to per-handle training state. This PR relaxes it again, which should restore compatibility for some reentrant AC plus FSDP cases.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Relax post-backward assert"


This assert was accidentally made stricter when transitioning from per-FSDP-instance training state to per-handle training state. This PR relaxes it again, which should restore compatibility for some reentrant AC plus FSDP cases.

[ghstack-poisoned]

awgu commented:
@pytorchbot rebase -s

Commit:
Update on "[FSDP] Relax post-backward assert"


This assert was accidentally made stricter when transitioning from per-FSDP-instance training state to per-handle training state. This PR relaxes it again, which should restore compatibility for some reentrant AC plus FSDP cases.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

awgu commented:
@pytorchbot merge
